titan-compiler/lexer.lua

```diff
@@ -0,0 +1,217 @@
+--
+-- This module exports lpeg patterns that can lex Titan source code.
+--
+-- The lexer follows the "longest match" rule and for any given input there
+-- will be at most one token that matches it. For example, lexer.LE matches
+-- the start of "<=bla" but lexer.LT doesn't. This should mean that you don't
+-- need to worry about what token comes first in a PEG ordered choice or about
+-- the lexer splitting words in the middle (such as "localx" being parsed as
+-- "local" "x").
+--
+-- The exported tokens are in ALLCAPS, to play nice with the "re"-style grammars
+-- from parser-gen: it can only refer to our tokens if we use alphabetical
+-- identifiers and it expects terminals to be uppercase and non-terminals to be
+-- lowercase.
+
+local lpeg = require 'lpeglabel'
+local P = lpeg.P
+local R = lpeg.R
+local S = lpeg.S
+local C = lpeg.C
+local Cmt = lpeg.Cmt
+
+local lexer = {}
+
+--
+-- Numbers
+--
+
+-- This very general pattern matches both integer and floating point numbers,
+-- in either decimal or hexadecimal notation, as well as a bunch of invalid
+-- stuff. We use tonumber in the end to find out which is which. This pattern
+-- is intentionally more general than the one that Lua uses to avoid allowing
+-- weird things such as `1337require`.
+local number_start = P(".")^-1 * R("09")
+local expo = S("EePp") * S("+-")^-1
+local possible_number = #number_start * (expo + R("09", "AZ", "az") + ".")^1
+lexer.NUMBER = Cmt(possible_number, function(_, i, s)
+    local n = tonumber(s)
+    if n then
+        return i, n
+    else
+        assert(false, "todo: nice syntax error for malformed number")
+    end
+end)
+
+--
+-- Strings
+--
+
+-- if s[i] points to a newline sequence, skip it
+local function skip_linebreak(s, i)
+    local c1, c2 = string.byte(s, i, i+1)
+    if     c1 == 10 and c2 == 13 then return i + 2
```

Comment:
We need to add a comment stating what is '10' and '13'.
Also, why do we need \r\n and \n\r?

Reply:
Could this be a if, elseif with two cases, and without the fancy alignment?